### ISSUE TITLE
Cannot get session in pages/api/graphql.ts

### ISSUE THREAD

**Issue by 543com** [apollo-server-integrations/apollo-server-integration-next#150]
I tried to add auth for /api/graphql
I found I cannot get session info in startServerAndCreateNextHandler
Below console.log will print null, so I cannot use session for auth.

```
// src/pages/api/graphql.ts
import { ApolloServer } from "@apollo/server";
import { startServerAndCreateNextHandler } from "@as-integrations/next";
import { getServerSession } from "next-auth/next";
import { authOptions } from "./auth/[...nextauth]";

import typeDefs from "@/graphql/schemas";
import resolvers from "@/graphql/resolvers";
import allowCors from "@/utils/cors";

const apolloServer = new ApolloServer({
  typeDefs,
  resolvers,
});

const handler = startServerAndCreateNextHandler(apolloServer, {
  context: async (req, res) => {
    const session = await getServerSession(req, res, authOptions);

    console.log("Session in api/graphql:", session);

    return { req, res, context };
  },
});

export default handler;
```

```
// src/pages/api/auth/[...nextauth].ts
import NextAuth, { User, Session, Profile, NextAuthOptions } from "next-auth";
import GoogleProvider from "next-auth/providers/google";
import { PrismaAdapter } from "@next-auth/prisma-adapter";

import prisma from "@/libs/prisma";

export const authOptions = {
  adapter: PrismaAdapter(prisma),

  providers: [
    GoogleProvider({
      clientId: process.env.GOOGLE_CLIENT_ID || "",
      clientSecret: process.env.GOOGLE_CLIENT_SECRET || "",
    }),
  ],

  secret: process.env.SECRET,

  callbacks: {
    async session({ session, user }: { session: Session; user: User }) {
      const prismaUser = await prisma.user.findUnique({
        where: { email: user.email! },
      });
      if (prismaUser) session.user.roles = prismaUser.roles;
      return session;
    },
  },
};

export default NextAuth(authOptions);
```

PS:  `getServerSession(req, res, authOptions)` is working in other api like below:

```
//src/pages/api/order/[orderId]/index.ts
import { NextApiRequest, NextApiResponse } from "next";

import { getServerSession } from "next-auth/next";
import { authOptions } from "@/pages/api/auth/[...nextauth]";

import prisma from "@/libs/prisma";

export default async function handle(
  req: NextApiRequest,
  res: NextApiResponse,
) {
  const session = await getServerSession(req, res, authOptions);

  console.log("session!:", session);

  if (!session || !session.user.roles.includes("ADMIN")) {
    return res.status(401).json({ message: "Unauthorized" });
  }

  const orderId = Array.isArray(req.query.id) ? req.query.id[0] : req.query.id;

  if (req.method === "GET") {
    const orderProducts = await prisma.orderProduct.findMany({
      where: { orderId: orderId },
      include: {
        product: true,
      },
    });
    // console.log(orderProducts);
    res.json(orderProducts);
  } else {
    res.status(405).json({ message: "Method not allowed" });
  }
}
```

**Comment by trevor-scheer** [comment]
@543com it's not obvious to me why this isn't working for you, can you share a reproduction? Have you done any debugging to dig into the issue?

**Comment by 543com** [comment]
how can I provide extra info for you ?

**Comment by trevor-scheer** [comment]
@543com if you can share a small repo that demonstrates the issue so that I can run it locally and debug it that would be great.

**Comment by jharlowuk** [comment]
Keenly interested in this thread. I've been playing around with this as well and am having the same issue.

**Comment by jharlowuk** [comment]
Ok, I was able to get it working in the `pages/api` directory, but it only works with `useQuery`, not the experimental query support for Next13. 

Here's my code, I am successfully getting session data for calls made from `pages/index.tsx` with `useQuery`. I suspect we may just have to wait for tooling for the `app` dir to mature.

```ts
export const getContext: ContextFunction<[NextApiRequest, NextApiResponse], GraphQLContext> = async (req, res) => {
  const session = await getServerSession(req, res, authOptions);
  return {
    req,
    dataSources: { prisma },
    userInfo: { session },
  };
};

```

**Comment by trevor-scheer** [comment]
Thanks for doing some digging @jharlowuk. Are you suggesting this is probably a next issue? Can you share a reproduction here? If it _is_ a next issue, maybe we can minimize it further and open an issue against their project?

**Comment by jharlowuk** [comment]
I'm not 100% certain, however, based on the debugging I did I concluded that the problem is with `@apollo/experimental-nextjs-app-support`. I followed [this tutorial](https://www.apollographql.com/blog/apollo-client/next-js/how-to-use-apollo-client-with-next-js-13/) and was able to get data from my ApolloServer in `app/api/graphql/route.ts` just fine, however, I think something is going on with the headers with that client that prevents the right `req` and `res` data being passed to the server.

I think this because:
- My data is otherwise coming back fine (for example if I use `ExampleQuery` that just sends `return "Hello" + ctx.user.name`, I get `"Hello undefined"` back on the client)
- All other `contextValues` are otherwise fine.
- `await getServerSession(authOptions)` works a-OK in the `page.tsx` file that I also invoke `await getClient()` in, so clearly the actual NextJS `Request` and `Response` data are being interpreted ok!

That's probably a not-perfect explanation. I'll try to find the time to put together a small repo to showcase but I unfortunately can't share the main repo I am building with so putting one together is around ~1 hour of extra work that I need to find somewhere! Also thanks for the new approach to AS integrations, I really like the change from 3->4.

**Comment by jharlowuk** [comment]
OK, I've done some more work and have managed to get it working (am receiving a `getServerSideSession` details correctly inside `context`)! I can confirm that this issue had nothing to do with this repo (sorry for creating a lot of noise here). I'm going to take some time to write a blog post because I think that with this working, Apollo + SSR + NextAuth + nexus + codegen has become an insanely nice DX. Will crosspost the writeup here.

**Comment by jharlowuk** [comment]
TLDR: @trevor-scheer I think it's pretty safe to close this issue, you just need to pass the necessary headers in the `links` array

**Comment by trevor-scheer** [comment]
Thanks @jharlowuk, looking forward to the writeup! Is there anything we can improve on the `@apollo/experimental-nextjs-app-support` side? I can pass this along to the maintainers if so.

I'm not sure if I missed it or if it's not written up above, but what is the workaround? It sounds like you were able to identify the issue and work past it, but I'm sure @543com (and any others) would be interested to know what the solve is.
